src/rdkafka_sasl_oauthbearer_oidc.c

```diff
@@ -0,0 +1,529 @@
+        char *decoded_payloads = NULL;
+
+        struct curl_slist *headers = NULL;
+
+        const char *token_url;
+        const char *extension_str;
+        const char *sub;
+
+        size_t post_fields_size;
+        size_t extension_cnt;
+        size_t extension_key_value_cnt;
+
+        char set_token_errstr[512];
+
+        char **extensions;
+        char **extension_key_value;
+
+        if (rd_kafka_terminating(rk))
+                return;
+
+        rd_kafka_oidc_build_headers(rk->rk_conf.sasl.oauthbearer.client_id,
+                                    rk->rk_conf.sasl.oauthbearer.client_secret,
+                                    &headers);
+
+        /* Build post fields */
+        post_fields_size = strlen("grant_type=client_credentials&scope=") +
+                           strlen(rk->rk_conf.sasl.oauthbearer.scope) + 1;
+        post_fields = rd_malloc(post_fields_size);
+        rd_snprintf(post_fields, post_fields_size,
+                    "grant_type=client_credentials&scope=%s",
+                    rk->rk_conf.sasl.oauthbearer.scope);
+
+        token_url = rk->rk_conf.sasl.oauthbearer.token_endpoint_url;
+
+        herr = rd_http_post_expect_json(rk, token_url, headers, post_fields,
+                                        post_fields_size, timeout_s, retry,
+                                        retry_ms, &json);
+
+        if (unlikely(herr != NULL)) {
+                rd_kafka_log(rk, LOG_ERR, "OIDC",
+                             "Failed to retrieve OIDC "
+                             "token from \"%s\": %s (%d)",
+                             token_url, herr->errstr, herr->code);
+                rd_kafka_oauthbearer_set_token_failure(rk, herr->errstr);
+                rd_http_error_destroy(herr);
+                goto done;
+        }
+
+        if (!cJSON_HasObjectItem(json, "access_token")) {
+                rd_kafka_oauthbearer_set_token_failure(
+                    rk,
+                    "Expected JSON response with "
+                    "\"access_token\" field");
+                goto done;
+        }
+
+        parsed_token = cJSON_GetObjectItem(json, "access_token");
+
+        jwt_token = cJSON_GetStringValue(parsed_token);
+        if (jwt_token == NULL) {
+                rd_kafka_oauthbearer_set_token_failure(
+                    rk,
+                    "Expected JSON "
+                    "response as a value string");
+                goto done;
+        }
+
+        decoded_payloads = rd_kafka_jwt_b64_decode(jwt_token);
+        payloads         = cJSON_Parse(decoded_payloads);
+
+        if (!cJSON_HasObjectItem(payloads, "exp")) {
+                rd_kafka_oauthbearer_set_token_failure(
+                    rk,
+                    "Expected JSON response with "
+                    "\"exp\" field");
+                goto done;
+        }
+        jwt_exp = cJSON_GetObjectItem(payloads, "exp");
+        exp     = cJSON_GetNumberValue(jwt_exp);
+        if (isnan(exp)) {
```

Comment:
Maybe we can make this even simpler and not rely on isnan by checking that exp is positive (e.g., greater than 0.1 or something)? We'll want exp to be in the future anyway.

Reply:
I used the isnan() because the cJSON_GetNumberValue will return NAN if the "exp" is not a number.

Reply:
Problem is there is no guarantee that isnan() in this file and in cJSON.c behaves the same way. One definition might be from the fallback define, while the other one is using a gcc builtin, all depending on include file order and stuff.  This makes it hard to guarantee that the outcome is identical on all platforms, compilers, architectures, etc, and why I'm urging a different approach, even though it might seem uglier.